src/test/java/org/opentripplanner/apis/gtfs/PatternByServiceDatesFilterTest.java

```diff
@@ -0,0 +1,200 @@
+      .withArrivalTimes("10:00 10:05")
+      .withDepartureTimes("10:00 10:05")
+      .build();
+    pattern.add(tt);
+    return pattern;
+  }
+
+  static List<Arguments> invalidRange() {
+    return List.of(
+      Arguments.of(parse("2024-05-02"), parse("2024-05-01")),
+      Arguments.of(parse("2024-05-03"), parse("2024-05-01"))
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("invalidRange")
+  void invalidRange(LocalDate start, LocalDate end) {
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> new PatternByServiceDatesFilter(EMPTY_SERVICE, start, end)
+    );
+  }
+
+  static List<Arguments> validRange() {
+    return List.of(
+      Arguments.of(parse("2024-05-02"), parse("2024-05-02")),
+      Arguments.of(parse("2024-05-02"), parse("2024-05-03")),
+      Arguments.of(null, parse("2024-05-03")),
+      Arguments.of(parse("2024-05-03"), null),
+      Arguments.of(null, null)
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("validRange")
+  void validRange(LocalDate start, LocalDate end) {
+    assertDoesNotThrow(() -> new PatternByServiceDatesFilter(EMPTY_SERVICE, start, end));
+  }
+
+  static List<Arguments> ranges() {
+    return List.of(
+      Arguments.of(null, null, NOT_REMOVED),
+      Arguments.of(null, parse("2024-05-03"), NOT_REMOVED),
+      Arguments.of(null, parse("2024-05-01"), NOT_REMOVED),
+      Arguments.of(parse("2024-05-03"), null, NOT_REMOVED),
+      Arguments.of(parse("2024-05-01"), null, NOT_REMOVED),
+      Arguments.of(parse("2024-05-02"), parse("2024-05-02"), REMOVED),
+      Arguments.of(parse("2024-05-02"), parse("2024-05-03"), REMOVED),
+      Arguments.of(parse("2025-01-01"), null, REMOVED),
+      Arguments.of(parse("2025-01-01"), parse("2025-01-02"), REMOVED),
+      Arguments.of(null, parse("2023-12-31"), REMOVED),
+      Arguments.of(parse("2023-12-31"), parse("2024-04-30"), REMOVED)
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("ranges")
+  void filterPatterns(LocalDate start, LocalDate end, FilterExpectation expectation) {
+    var service = mockService();
+    var filter = new PatternByServiceDatesFilter(service, start, end);
+
+    var filterInput = List.of(PATTERN_1);
+    var filterOutput = filter.filterPatterns(filterInput);
+
+    if (expectation == NOT_REMOVED) {
+      assertEquals(filterOutput, filterInput);
+    } else {
+      assertEquals(List.of(), filterOutput);
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("ranges")
+  void filterRoutes(LocalDate start, LocalDate end, FilterExpectation expectation) {
+    var service = mockService();
+    var filter = new PatternByServiceDatesFilter(service, start, end);
+
+    var filterInput = List.of(ROUTE_1);
+    var filterOutput = filter.filterRoutes(filterInput.stream());
+
+    if (expectation == NOT_REMOVED) {
+      assertEquals(filterOutput, filterInput);
+    } else {
+      assertEquals(List.of(), filterOutput);
+    }
+  }
+
+  private static TransitService mockService() {
```

Comment:
There isn't a simple way to create an actual service without mocking?

Reply:
I instantiated a calendar service instead: 3c503de1e24ad35d922ac1d16b594799102fcec9

Do you prefer that?

Reply:
I think it's at least to the better direction but we can discuss this in the dev meeting.

Reply:
Refactored in https://github.com/opentripplanner/OpenTripPlanner/pull/5869/commits/b618e716e073516f93beee42f4ee8916557815a3